### PR TITLE
spdx-utils: Update the SPDX license list to version 3.10

### PR DESCRIPTION
**Pull request by sschuberth**
Signed-off-by: Sebastian Schuberth <sebastian.schuberth@bosch.io>

### CODE REVIEW

**Review by mnonnenmacher — APPROVED** [review]

File: spdx-utils/src/main/kotlin/SpdxSimpleLicenseMapping.kt [hunk 1]
```diff
@@ -47,8 +47,7 @@ object SpdxSimpleLicenseMapping {
      */
     private val deprecatedLicenseIds by lazy {
         val resource = javaClass.getResource("/deprecated-license-mapping.yml")
-        val map = yamlMapper.readValue<Map<String, SpdxLicense>>(resource)
-        map.mapValues { (_, v) -> v.toExpression() }
+        yamlMapper.readValue<Map<String, SpdxSingleLicenseExpression>>(resource)
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxSimpleLicenseMapping.kt:50]: Shouldn't that be `SpdxLicenseIdExpression`? That's the type it had before.

File: spdx-utils/src/main/kotlin/SpdxLicenseException.kt [hunk 2]
```diff
@@ -88,6 +90,8 @@ enum class SpdxLicenseException(
         /**
          * Return the enum value for the given [id], or null if it is no SPDX license exception id.
          */
+        @JsonCreator(mode = JsonCreator.Mode.DELEGATING)
```
Comment (mnonnenmacher) [inline spdx-utils/src/main/kotlin/SpdxLicenseException.kt:93]: Why is this necessary? I have problems understanding where it is used. Maybe explain it in the commit message.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: spdx-utils/src/main/kotlin/SpdxSimpleLicenseMapping.kt (hunk 1 above)
Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxSimpleLicenseMapping.kt:50]: You're basically right, but unfortunately this does not work as the default constructor of `SpdxLicenseIdExpression` sets `orLaterVersion` to `false`, which means it fails to deserialize e.g. an `id` of "AGPL-1.0-or-later", which would require `orLaterVersion` to be set to `true`. The direct parent class `SpdxSimpleExpression` also cannot be used as it has no constructor. So I'm using the grand-parent class `SpdxSingleLicenseExpression`, which also is a preparation towards merging the mapping for licenses and excpetions, which I was planning to do.

File: spdx-utils/src/main/kotlin/SpdxLicenseException.kt (hunk 2 above)
Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxLicenseException.kt:93]: Without this, the license / exception enums would require their *names* to be used instead of their *ids* for deserialization, and the point of this change is to not use ORT-specific enum names in the `yml`-mapping files anymore, so these files could be used by other parties, too.

Comment (mnonnenmacher) [reply spdx-utils/src/main/kotlin/SpdxLicenseException.kt:93]: Yes, but my problem is still to find whereis this constructor used for deserialization? I guess the one for `SpdxLicense` is used here:
https://github.com/oss-review-toolkit/ort/blob/d4f0cb598cc9af4402be6c4f5ee599a6c8136bea/spdx-utils/src/main/kotlin/SpdxSimpleLicenseMapping.kt#L34-L37

But where is the one for exceptions used?

Comment (sschuberth) [reply spdx-utils/src/main/kotlin/SpdxLicenseException.kt:93]: > I guess the one for `SpdxLicense` is used here:

Correct.

> But where is the one for exceptions used?

Nowhere. But since both files are generated from the same template by a Gradle task, they both contain these lines. But having them also in `SpdxLicenseException.kt` shouldn't hurt, and is more consistent.

**Review by mnonnenmacher — APPROVED** [review]
